Add new MVA tau ID for phase II

swozniewski opened the pull request:
#### PR description:

Adds new MVA tau ID for phase II to be run on miniAOD.
Presentation: https://indico.cern.ch/event/884712/contributions/3736447/attachments/1981798/3300601/TauIDPhase2_05022020.pdf
Adds a dedicated test analyzer.
Adds production of the ID to the miniAOD sequence triggered by a phase II modifier.

Test command to run the miniAOD sequence with phase II modifier:
`cmsDriver.py miniAOD2026 --filein root://eoscms.cern.ch//eos/cms/store/relval/CMSSW_11_1_0_pre8/RelValTTbar_14TeV/GEN-SIM-RECO/PU25ns_110X_mcRun4_realistic_v3_2026D49PU200-v1/20000/FF37C15C-1A4E-794E-B39F-B9B5D1860DAF.root --fileout file:MiniAOD2026.root --mc --eventcontent MINIAODSIM --runUnscheduled --datatier MINIAODSIM --conditions auto:phase2_realistic_T14 --step PAT --nThreads 4 --geometry Extended2026D49 --scenario pp --era Phase2C8 -n 2`

#### PR validation:

- Test command shown above successfully runs miniAOD creating the phaseII ID
- compiles
- code checks pass
- format checks pass
- unit tests pass
- limited matrix tests pass

Commit:
Phase 2 tau iso training

Commit:
eveything up to running in official cmssw

Commit:
remove minor changes which make the PR diff ugly

Commit:
eveything up to running in official cmssw, 2nd try

Commit:
make diff nicer

Commit:
diff shows nothing...

Commit:
add training to miniAOD_tools.py

Commit:
std::abs, remove local db commented section + cosmetics

Commit:
use packedCand for track

Commit:
std::abs, newlines

Commit:
remove self from runTauIdMVA.py for Phase2 entry

Commit:
fix capitalzation of ID name

Commit:
apply code format fixes

Commit:
fix phase II modification of miniAOD workflow

Commit:
simplify label for phase II MVA ID

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30341/16354

- This PR adds an extra 104KB to repository

- Found files with invalid states:
  - RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
    - Added: 65825d4b19a383e189472197bb86851269ad88bb
    - Deleted: 5fae4d0ced957aafd8cbe2e33d23208ae7b9ca45


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py  modified in PR(s): #30237
  - File  RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc  modified in PR(s): #29935

cmsbuild commented:
A new Pull Request was created by  @swozniewski  for master.

It involves the following packages:

PhysicsTools/PatAlgos
RecoTauTag/Configuration
RecoTauTag/RecoTau


@perrotta, @cmsbuild, @santocch, @slava77 can you please review it and eventually sign? Thanks.
@rappoccio, @gouskos, @hatakeyamak, @emilbols, @peruzzim, @seemasharmafnal, @mmarionncern, @ahinzmann, @smoortga, @jdolen, @ferencek, @jdamgov, @nhanvtran, @gkasieczka, @schoef, @andrzejnovak, @clelange, @riga, @JyothsnaKomaragiri, @mverzett, @gpetruc, @mariadalfonso this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 58905349025f7819fe7ab73e8e6024261deed07d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/7282/summary.html
CMSSW: CMSSW_11_2_X_2020-06-23-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/7282/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 7 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778811
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778757
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

santocch commented:
+1

perrotta commented:
assign upgrade

cmsbuild commented:
New categories assigned: upgrade

@kpedro88 you have been requested to review this Pull request/Issue and eventually sign? Thanks

kpedro88 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -339,6 +340,8 @@ def miniAOD_customizeCommon(process):
         toKeep = ['deepTau2017v2p1']
     )
+    from Configuration.Eras.Modifier_phase2_common_cff import phase2_common #Phase2 Tau MVA
+    phase2_common.toModify(tauIdEmbedder.toKeep, func=lambda t:t.append('newDMPhase2v1')) #Phase2 Tau MVA
```
do we need to keep the 2017 tau ID for phase 2, or can we just store the new one?

swozniewski replied:
yes, we would like to keep it for comparison.

kpedro88 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -348,6 +351,10 @@ def miniAOD_customizeCommon(process):
     )
     process.deepTauIDTask = cms.Task(process.deepTau2017v2p1, process.slimmedTaus)
     task.add(process.deepTauIDTask)
+    if 'newDMPhase2v1' in tauIdEmbedder.toKeep:
```
it would be preferable to do this with the `phase2_common` modifier (using `toModify()`, `toReplaceWith()`, etc.)

swozniewski replied:
At least for the task.add we would need a lambda in the modifier. We thought it's more compact this way and the modules should be run if this ID is 'to be kept' whether it is due to the phase2 modification or sth else. If you still prefer the modifier, please confirm and I will switch back.

kpedro88 replied:
You don't, actually. You can make an object called `_deepTauIDTaskPhase2` as a clone of `deepTauIDTask`, add the additional module, and then swap them using `toReplaceWith()`. (The leading underscore in the name ensures that it won't get attached to the process otherwise.)

swozniewski replied:
Thanks, I was not aware of this possibility. I will try

swozniewski replied:
It seems like `Task` has no method `clone()`. How did you mean to clone it? On the other hand it's maybe still better to do it via a lambda. This would have one line less if I understand your plan with the cloning correctly.

kpedro88 replied:
My mistake, for `Task` you have to use `copy()`, not `clone()`

mbluj replied:
@swozniewski you can find examples a few lines below :)

swozniewski replied:
thanks. Right, I just realized that :)

kpedro88 commented on RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc:
```diff
@@ -120,6 +122,8 @@ namespace reco {
                  mvaOpt_ == kDBoldDMwLTwGJ || mvaOpt_ == kDBnewDMwLTwGJ)
           mvaInput_ = new float[23];
+        else if (mvaOpt_ == kDBnewDMwLTwGJPhase2)
+          mvaInput_ = new float[30];
```
outside the scope of this PR, but in the future, this should move to using smart pointers (avoid manual memory management) and possibly `std::array`

swozniewski replied:
thanks for the comment! We will note that as a clean up ToDo. (@mbluj also for your interest)

kpedro88 commented on RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:
```diff
@@ -235,6 +235,113 @@ namespace reco {
+          float sigCands_pt = 0.;
+          float sigCands_dr, sigCands_deta, sigCands_dphi;
+          sigCands_dr = sigCands_deta = sigCands_dphi = 0.;
+          for (auto& j : tau.signalGammaCands()) {
```
const auto&

kpedro88 commented on RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:
```diff
@@ -235,6 +235,113 @@ namespace reco {
+          float isoCands_pt = 0.;
+          float isoCands_dr, isoCands_deta, isoCands_dphi;
+          isoCands_dr = isoCands_deta = isoCands_dphi = 0.;
+          for (auto& j : tau.isolationGammaCands()) {
```
const auto&

kpedro88 commented on RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:
```diff
@@ -235,6 +235,113 @@ namespace reco {
+          float thetaGJmax, thetaGJ;
+          thetaGJmax = thetaGJ = 0.;
+          if (decayDistMag > 0.) {
+            const float mTau = 1.77686;
```
define this constant in a more generic location (it does not depend on any conditional statement)

kpedro88 commented on RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.cc:
```diff
@@ -0,0 +1,191 @@
+#include "TH2D.h"
+#include "TGraphAsymmErrors.h"
+
+class rerunMVAIsolationOnMiniAOD_Phase2 : public edm::EDAnalyzer {
```
legacy analyzers should not be used
this should be `edm::one::EDAnalyzer<edm::one::SharedResources>` since it uses `TFileService`

swozniewski commented:
@kpedro88 thank you for your review! I gave dedicated responses to some comments and will provide a follow up commit for the others. Please let me know about your opinion concerning the 'if vs modifier'-case above.

Commit:
apply review comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

swozniewski commented:
Ok, I think I addressed your comments. I applied one additional change `By...` -> `by...` in order to fix the analyzer which was not changed consistently in one of the previou commits.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30341/16882

- This PR adds an extra 104KB to repository

- Found files with invalid states:
  - RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
    - Added: 7c62ecfad2d131829d6fb0b62fdc3af5f44fc89b
    - Deleted: 253c0347386a1ac91b5e48ba442164b0dfd5d8e3


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoTauTag/RecoTau/plugins/PATTauDiscriminationByMVAIsolationRun2.cc  modified in PR(s): #29935
  - File  RecoTauTag/RecoTau/python/tools/runTauIdMVA.py  modified in PR(s): #30581
  - File  RecoTauTag/RecoTau/test/BuildFile.xml  modified in PR(s): #30535

cmsbuild commented:
Pull request #30341 was updated. @perrotta, @kpedro88, @cmsbuild, @slava77, @jpata, @santocch can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: b57c24fe5382591b84a7d164548f2e544691687c

CMSSW: CMSSW_11_2_X_2020-07-09-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/7826/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
136.88811 step2
<pre>runTheMatrix-results/136.88811_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL/step2_RunJetHT2018D_reminiaodUL+RunJetHT2018D_reminiaodUL+REMINIAOD_data2018UL+HARVEST2018_REMINIAOD_data2018UL.log

</pre>136.8311 step2
<pre>runTheMatrix-results/136.8311_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017/step2_RunJetHT2017F_reminiaod+RunJetHT2017F_reminiaod+REMINIAOD_data2017+HARVEST2017_REMINIAOD_data2017.log

</pre>136.7611 step2
<pre>runTheMatrix-results/136.7611_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM/step2_RunJetHT2016E_reminiaod+RunJetHT2016E_reminiaod+REMINIAOD_data2016_HIPM+HARVESTDR2_REMINIAOD_data2016_HIPM.log

</pre>4.53 step3
<pre>runTheMatrix-results/4.53_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT/step3_RunPhoton2012B+RunPhoton2012B+HLTD+RECODR1reHLT+HARVESTDR1reHLT.log

</pre>158.01 step2
<pre>runTheMatrix-results/158.01_HydjetQ_reminiaodPbPb2018_INPUT+HydjetQ_reminiaodPbPb2018_INPUT+REMINIAODHI2018PPRECO+HARVESTHI2018PPRECOMINIAOD/step2_HydjetQ_reminiaodPbPb2018_INPUT+HydjetQ_reminiaodPbPb2018_INPUT+REMINIAODHI2018PPRECO+HARVESTHI2018PPRECOMINIAOD.log

</pre>136.731 step3
<pre>runTheMatrix-results/136.731_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2/step3_RunSinglePh2016B+RunSinglePh2016B+HLTDR2_2016+RECODR2_2016reHLT_skimSinglePh_HIPM+HARVESTDR2.log

</pre>136.793 step3
<pre>runTheMatrix-results/136.793_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017/step3_RunDoubleEG2017C+RunDoubleEG2017C+HLTDR2_2017+RECODR2_2017reHLT_skimDoubleEG_Prompt+HARVEST2017.log

</pre>136.874 step3
<pre>runTheMatrix-results/136.874_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM/step3_RunEGamma2018C+RunEGamma2018C+HLTDR2_2018+RECODR2_2018reHLT_skimEGamma_Offline_L1TEgDQM+HARVEST2018_L1TEgDQM.log

</pre>135.4 step3
<pre>runTheMatrix-results/135.4_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS/step3_ZEE_13+ZEEFS_13+HARVESTUP15FS+MINIAODMCUP15FS.log

</pre>1306.0 step3
<pre>runTheMatrix-results/1306.0_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15/step3_SingleMuPt1_UP15+SingleMuPt1_UP15+DIGIUP15+RECOUP15+HARVESTUP15.log

</pre>1330.0 step3
<pre>runTheMatrix-results/1330.0_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15/step3_ZMM_13+ZMM_13+DIGIUP15+RECOUP15_L1TMuDQM+HARVESTUP15_L1TMuDQM+NANOUP15.log

</pre>9.0 step3
<pre>runTheMatrix-results/9.0_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST/step3_Higgs200ChargedTaus+Higgs200ChargedTaus+DIGI+RECO+HARVEST.log

</pre>25.0 step3
<pre>runTheMatrix-results/25.0_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT/step3_TTbar+TTbar+DIGI+RECOAlCaCalo+HARVEST+ALCATT.log

</pre>158.0 step3
<pre>runTheMatrix-results/158.0_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO/step3_HydjetQ_B12_5020GeV_2018_ppReco+HydjetQ_B12_5020GeV_2018_ppReco+DIGIHI2018PPRECO+RECOHI2018PPRECO+ALCARECOHI2018PPRECO+HARVESTHI2018PPRECO.log

</pre>10042.0 step3
<pre>runTheMatrix-results/10042.0_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017/step3_ZMM_13+ZMM_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017.log

</pre>10024.0 step3
<pre>runTheMatrix-results/10024.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017_GenSimFull+DigiFull_2017+RecoFull_2017+HARVESTFull_2017+ALCAFull_2017+NanoFull_2017.log

</pre>25202.0 step3
<pre>runTheMatrix-results/25202.0_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25/step3_TTbar_13+TTbar_13+DIGIUP15_PU25+RECOUP15_PU25+HARVESTUP15_PU25+NANOUP15_PU25.log

</pre>10824.0 step3
<pre>runTheMatrix-results/10824.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018+NanoFull_2018/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2018_GenSimFull+DigiFull_2018+RecoFull_2018+HARVESTFull_2018+ALCAFull_2018+NanoFull_2018.log

</pre>11634.0 step3
<pre>runTheMatrix-results/11634.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021/step3_TTbar_14TeV+TTbar_14TeV_TuneCP5_2021_GenSimFull+DigiFull_2021+RecoFull_2021+HARVESTFull_2021+ALCAFull_2021.log

</pre>12434.0 step3
<pre>runTheMatrix-results/12434.0_TTbar_14TeV+TTbar_14TeV_TuneCP5_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023/step3_TTbar_14TeV+TTbar_14TeV_TuneCP5_2023_GenSimFull+DigiFull_2023+RecoFull_2023+HARVESTFull_2023+ALCAFull_2023.log

</pre>10224.0 step3
<pre>runTheMatrix-results/10224.0_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU+NanoFull_2017PU/step3_TTbar_13+TTbar_13TeV_TuneCUETP8M1_2017PU_GenSimFull+DigiFullPU_2017PU+RecoFullPU_2017PU+HARVESTFullPU_2017PU+NanoFull_2017PU.log

</pre>250202.181 step4
<pre>runTheMatrix-results/250202.181_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25/step4_TTbar_13UP18+TTbar_13UP18+PREMIXUP18_PU25+DIGIPRMXLOCALUP18_PU25+RECOPRMXUP18_PU25+HARVESTUP18_PU25.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

swozniewski commented:
Sorry, I forgot about this problem. This was actually also fixed by using the if-condition instead of modifiers. E.g. `process.rerunDiscriminationByIsolationMVADBnewDMwLTPhase2raw` needs to be modified for phase 2, but it is only created by the tauIdEmbedder if modified for phase2, causing a crash for non-phase2 tasks. Do you have an idea how this could be solved with modifiers? Or should I switch back to the `if`?

kpedro88 commented on PhysicsTools/PatAlgos/python/slimming/miniAOD_tools.py:
```diff
@@ -369,6 +372,12 @@ def miniAOD_customizeCommon(process):
     process.deepTauIDTask = cms.Task(process.deepTau2017v2p1, process.slimmedTaus)
     task.add(process.deepTauIDTask)
 
+    phase2_common.toModify(process.rerunDiscriminationByIsolationMVADBnewDMwLTPhase2raw, PATTauProducer=_noUpdatedTauName)
```
based on https://github.com/cms-sw/cmssw/pull/30341#issuecomment-656327868, I propose:
```python
for _prod in ["rerunDiscriminationByIsolationMVADBnewDMwLTPhase2raw","rerunDiscriminationByIsolationMVADBnewDMwLTPhase2"]:
    if hasattr(process,_prod): phase2_common.toModify(getattr(process,prod), PATTauProducer=_noUpdatedTauName)
```
@makortel do you have any better ideas?

makortel replied:
That looks fine for a customize function. The approach with cff fragments is that all modules are defined for all eras, and the ones to be actually run are handled by customizing Tasks/Sequences.

swozniewski replied:
Is that really better than the original if-statement? Maybe I miss an advantage of having the modifier included, but at the moment this looks more confusing to me.

swozniewski replied:
@kpedro88 please confirm your preferred solution.

kpedro88 replied:
Either is fine

swozniewski replied:
Ok, I switched back to the previous

cmsbuild commented:
The code-checks are being triggered in jenkins.

Commit:
remove added empty line

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30341/17039

- This PR adds an extra 104KB to repository

- Found files with invalid states:
  - fatal: bad object 55a9b81bd873dd53887ed8900f41c2425f7ae448


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoTauTag/RecoTau/python/tools/runTauIdMVA.py  modified in PR(s): #30581

cmsbuild commented:
Pull request #30341 was updated. @perrotta, @kpedro88, @cmsbuild, @slava77, @jpata, @santocch can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30341/17040

- This PR adds an extra 100KB to repository

- Found files with invalid states:
  - RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
    - Added: 7c62ecfad2d131829d6fb0b62fdc3af5f44fc89b
    - Deleted: 253c0347386a1ac91b5e48ba442164b0dfd5d8e3


- There are other open Pull requests which might conflict with changes you have proposed:
  - File  RecoTauTag/RecoTau/python/tools/runTauIdMVA.py  modified in PR(s): #30581

cmsbuild commented:
+1
Tested at: 55a9b81bd873dd53887ed8900f41c2425f7ae448
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/7981/summary.html
CMSSW: CMSSW_11_2_X_2020-07-14-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/7981/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2612943
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2612894
 - DQMHistoTests: Total skipped: 48
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 146 log files, 17 edm output root files, 35 DQM output files

perrotta commented on RecoTauTag/RecoTau/test/rerunMVAIsolationOnMiniAOD_Phase2.py:
```diff
@@ -0,0 +1,106 @@
+
+from RecoTauTag.RecoTau.PATTauDiscriminationByMVAIsolationRun2_cff import *
+process.rerunDiscriminationByIsolationMVADBnewDMwLTPhase2raw = patDiscriminationByIsolationMVArun2v1raw.clone(
+    PATTauProducer = cms.InputTag('slimmedTaus'),
```
This is a test config, but the same suggestion of dropping the type specs can still hold (here and below)

perrotta commented on RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:
```diff
@@ -863,6 +864,57 @@ def runTauID(self):
+            def tauIDMVAinputs(module, wp):
+                return cms.PSet(inputTag = cms.InputTag(module), workingPointIndex = cms.int32(-1 if wp=="raw" else -2 if wp=="category" else getattr(self.process, module).workingPoints.index(wp)))
+            self.process.rerunDiscriminationByIsolationMVADBnewDMwLTPhase2raw = patDiscriminationByIsolationMVArun2v1raw.clone(
+                PATTauProducer = cms.InputTag('slimmedTaus'),
```
Please drop type specification for the parameters modified in cloned configs, for a safer syntax (see item 4 in https://twiki.cern.ch/twiki/bin/view/CMS/OpenRecoTasks#Review_and_optimization_of_t_AN1)

perrotta commented on RecoTauTag/RecoTau/python/tools/runTauIdMVA.py:
```diff
@@ -863,6 +864,57 @@ def runTauID(self):
+            self.process.rerunDiscriminationByIsolationMVADBnewDMwLTPhase2raw = patDiscriminationByIsolationMVArun2v1raw.clone(
+                PATTauProducer = cms.InputTag('slimmedTaus'),
+                Prediscriminants = noPrediscriminants,
+                loadMVAfromDB = cms.bool(True),
```
same here (and below)

perrotta commented:
Checking on 100 events with wf 23234.0, this PR adds 125 B to slimmedTaus, which is the 0.3% of thw whole miniAOD event content. Not a huge increase, and fairly acceptable: just keep in mind to remove the 2017 tau ID from Phase2 as soon as you don't need to have them stored for comparison.

```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
   1906.7 ->      2031.9        125      6.4   0.30     patTaus_slimmedTaus__RECO.
-------------------------------------------------------------
    41264 ->       41389        125             0.3     ALL BRANCHES
```

perrotta commented:
@swozniewski please clarify on the status of this PR
Thank you.

swozniewski commented:
I did not yet look into your latest set of comments. I'm actually on vacation this week, so I don't want to make promises, but my plan was to however address the comments today or tomorrow in order to not further delay this PR.

perrotta commented:
> I did not yet look into your latest set of comments. I'm actually on vacation this week, so I don't want to make promises, but my plan was to however address the comments today or tomorrow in order to not further delay this PR.

Thank you @swozniewski 
I just wanted to check the status and verify that the comments got noticed,:enjoy your vacation!

Commit:
followup review comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

swozniewski commented:
Thanks! And thank you for the review. These are good points. I pushed a corresponding commit.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30341/17257

- This PR adds an extra 116KB to repository

- Found files with invalid states:
  - RecoTauTag/RecoTau/plugins/rerunMVAIsolationOnMiniAOD_Phase2.cc:
    - Added: 7c62ecfad2d131829d6fb0b62fdc3af5f44fc89b
    - Deleted: 253c0347386a1ac91b5e48ba442164b0dfd5d8e3

cmsbuild commented:
Pull request #30341 was updated. @perrotta, @kpedro88, @cmsbuild, @slava77, @jpata, @santocch can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 1fce3932b94eeb7380a5a335953784290d8e2b8f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/8233/summary.html
CMSSW: CMSSW_11_2_X_2020-07-22-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-e868b2/8233/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2526188
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2526134
 - DQMHistoTests: Total skipped: 47
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 144 log files, 17 edm output root files, 34 DQM output files

perrotta commented:
+1
- Specific TauID for Phase2 are defined, and their output stored in Phase2 miniAOD
- This increases the Phase2 miniAOD size by O(100 B), which can be recovered as soon as the contemporary storage of "2017" tau IDs kept for comparison can be removed from the event content
- Jenkins tests pass and only show the above differences in the Phase2 miniAOD outputs

kpedro88 commented:
+upgrade

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
+1

slava77 commented:
after this PR was merged we have some non-reproducible behavior in phase-2 slimmedTaus plots (it starts from CMSSW_11_2_X_2020-07-25-1100 IB)

slava77 commented on RecoTauTag/RecoTau/src/PFRecoTauClusterVariables.cc:
```diff
@@ -235,6 +235,111 @@ namespace reco {
+          mvaInput[3] = neutralIsoPtSum;               //tauID("neutralIsoPtSum");
+          mvaInput[4] = puCorrPtSum;                   //tauID("puCorrPtSum");
+          mvaInput[5] = photonPtSumOutsideSignalCone;  //tauID("photonPtSumOutsideSignalCone");
+          mvaInput[7] = tauDecayMode;                  //tau.decayMode();
```
I suppose this was meant to be 6.
I guess this is the origin of random results in jenkins tests

kpedro88 replied:
Another example of the need for #29818...
